pkg/reconciler/revision/reconcile_resources.go

```diff
@@ -94,29 +93,47 @@ func (c *Reconciler) reconcileDeployment(ctx context.Context, rev *v1.Revision)
 				}
 			}
 
+			// if a Pod is terminating already, we do not care it being not ready because this is expected to be the case
+			if pod.DeletionTimestamp != nil {
+				continue
+			}
+
+			// if a Pod is ready, then do not check it. The fact it is ready may not yet be reflected on the Deployment status.
```

Comment:
Previous code covered that already as if a pod is ready `if t := status.LastTerminationState.Terminated;` would never be true for a container.
Not sure how often this inconsistency will happen tbh to be sure about the benefits of the optimization.

Reply:
The scenario I have in mind is a following:

* a Pod is created and runs fine
* at some point the user-container goes out of memory and is restarted, from now on it has `LastTerminationState.Terminated` set.
* The container finshes restarting and becomes ready. Though, this may not yet be reflected in the deployment status due to the asynchronous nature of the controller.
* now this loop runs

The Pod will be Ready but if we never check this we would still iterate the container statuses. And there `status.LastTerminationState.Terminated` is not `nil`.

I could probably still get rid of it, if I change the loop to:

```go
// Iterate the container status and report the first failure
			for _, status := range pod.Status.ContainerStatuses {
				if status.Name == resources.QueueContainerName || status.Ready {
					continue
				}

				...
			}
```

Note the added `|| status.Ready` at the beginning. This might actually be nicer because we then generally do not report a potentially days-ago failure on a ready container (which would matter if on a single pod multiple containers have restarts). Wdyt?

Reply:
If a pod is ready all containers are running. Let's say we are keep crashing/restarting. The question is if we ever are going to reconcile during the time that the pod is down and capture it.

I think if we wanted to detect this reliably without caring how old a failure is, we would need a pod watcher, check [this](https://www.cncf.io/blog/2022/08/23/tracking-container-restarts-and-termination-events-in-kubernetes/) that explains it. This way we would now if we have old issues or that if some container is keep restarting.

An implicit way to detect the same I suspect is to check if we are in waiting state when `status.LastTerminationState.Terminated != nil`.
Before a container is restarted there is [a backoff time](https://sysdig.com/blog/debug-kubernetes-crashloopbackoff/) and container is in waiting stay with reason = "CrashLoopBackOff".